### ISSUE TITLE
TypeError: on_train_batch_start() takes 3 positional arguments but 4 were given

### ISSUE THREAD

**Issue by cxhagd** [Lightning-AI/pytorch-lightning#20107]
a solved problem:

Hi, I just started using Lightning and I've been running into this bug lately.

Here's my code:

```
from typing import Type, Mapping, Callable, Optional, Union, Tuple, List
import pytorch_lightning as pl
import torch
from torch import Tensor
from torch_geometric.data.storage import recursive_apply
from torchmetrics import Metric

from tsl.predictors import Predictor

class Imputer(Predictor):
    ...
    def on_train_batch_start(self, batch, batch_idx: int,
                                 unused: Optional[int] = 0) -> None:
        r"""For every training batch, randomly mask out value with probability
        :math:`p = \texttt{self.whiten\_prob}`. Then, whiten missing values in
         :obj:`batch.input.x`"""
        super(Imputer, self).on_train_batch_start(batch, batch_idx, unused)
        # randomly mask out value with probability p = whiten_prob
        batch.original_mask = mask = batch.input.mask
        p = self.whiten_prob
        if isinstance(p, Tensor):
            p_size = [mask.size(0)] + [1] * (mask.ndim - 1)
            p = p[torch.randint(len(p), p_size)].to(device=mask.device)
        whiten_mask = torch.rand(mask.size(), device=mask.device) > p
        batch.input.mask = mask & whiten_mask
        # whiten missing values
        if 'x' in batch.input:
            batch.input.x = batch.input.x * batch.input.mask
    ...
```
Here's my traceback:

```
Traceback (most recent call last):
  File "./experiments/run_imputation.py", line 313, in <module>
    run_experiment(args)
  File "./experiments/run_imputation.py", line 286, in run_experiment
    trainer.fit(imputer,
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 543, in fit
    call._call_and_handle_interrupt(
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 579, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 986, in _run
    results = self._run_stage()
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1030, in _run_stage
    self.fit_loop.run()
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 205, in run
    self.advance()
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 363, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/training_epoch_loop.py", line 140, in run
    self.advance(data_fetcher)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/training_epoch_loop.py", line 234, in advance
    response = call._call_lightning_module_hook(trainer, "on_train_batch_start", batch, batch_idx)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 159, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/tsl/imputers/imputer.py", line 134, in on_train_batch_start
    super(Imputer, self).on_train_batch_start(batch, batch_idx, unused)
TypeError: on_train_batch_start() takes 3 positional arguments but 4 were given

```

I asked ChatGPT to solve this problem, and it suggested changing:

def on_train_batch_start(self, batch, batch_idx: int, 
                         unused: Optional[int] = 0) -> None:
to:

def on_train_batch_start(self, batch, batch_idx: int, dataloader_idx=None,
                         unused: Optional[int] = 0) -> None:
But this caused a new issue. Here's the new traceback:

Traceback (most recent call last):
  File "./experiments/run_imputation.py", line 313, in <module>
    run_experiment(args)
  File "./experiments/run_imputation.py", line 286, in run_experiment
    trainer.fit(imputer,
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 543, in fit
    call._call_and_handle_interrupt(
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 579, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 986, in _run
    results = self._run_stage()
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1030, in _run_stage
    self.fit_loop.run()
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 205, in run
    self.advance()
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 363, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/training_epoch_loop.py", line 140, in run
    self.advance(data_fetcher)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/loops/training_epoch_loop.py", line 234, in advance
    response = call._call_lightning_module_hook(trainer, "on_train_batch_start", batch, batch_idx)
  File "/root/anaconda3/envs/spin/lib/python3.8/site-packages/pytorch_lightning/trainer/call.py", line 159, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
TypeError: on_train_batch_start() missing 1 required positional argument: 'dataloader_idx'
I'm confused! Could you please help me resolve this issue? Thank you!

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

Oh! I solved it!
The super(Imputer, self).on_train_batch_start(batch, batch_idx, unused)
was calling the parent class's function instead of the one defined in the subclass.
By tracing the inheritance hierarchy, I found that the on_train_batch_start function in the parent class's parent class's parent is defined as follows:
def on_train_batch_start(self, batch: Any, batch_idx: int) -> Optional[int]:
This means I just needed to remove the unused parameter in the call to resolve the issue.

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 1.5.0):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

**Comment by cxhagd** [comment]
Python: 3.12.4
PyTorch: 2.3.1+cu121
PyTorch Lightning: 2.3.3

**Comment by cxhagd** [comment]
Oh! I solved it! 
The `super(Imputer, self).on_train_batch_start(batch, batch_idx, unused)` 
was calling the parent class's function instead of the one defined in the subclass.
By tracing the inheritance hierarchy, I found that the `on_train_batch_start` function in the parent class's parent class's parent is defined as follows:
`def on_train_batch_start(self, batch: Any, batch_idx: int) -> Optional[int]:`
This means I just needed to remove the `unused` parameter in the call to resolve the issue.
